enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+---
+title: microshift-pluggable-core-components
+authors:
+  - majopela
+reviewers:
+  - "@copejon, MicroShift contributor"
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@dhellmann, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@zshi-redhat, MicroShift "
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2022-12-19
+last-updated: 2022-12-19
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-599
+---
+
+# MicroShift Pluggable Core Components
+
+## Summary
+
+This enhancement proposes making some of the core components like CNI, CSI or router
+optional and pluggable.
+
+In the enhancement proposal we will describe an architecture which would allow
+customers to install individual compomnents as rpm packages.
+
+
+## Motivation
+
+The principal factor that gave birth to the idea of MicroShift was the need to bring
+the OpenShift deployment model into the Edge where the power, memory and CPU budgets
+are reduced.
+
+While the default configuration makes sense to most of our customers in the Device
+Edge area, still our current CSI, CNI, and router components use 35% of the memory
+footprint and a significative CPU and power footprint. Making those components
+pluggable would allow customers who don't require advanced network features,
+persistent volumes, or ingress/routes to disable or switch those technologies
+and reduce the memory/cpu and disk/over-the-wire footprint.
+
+
+### User Stories
+
+An application developer needs the whole stack of components, creates an image
+that contains:
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
```

Comment:
Good point

We should remove the dependency from microshift.service and create a dependency from microshift-ovs-init service saying that it's wanted by microshift, we can write this in microshift-ovs-init.service:

Before=microshift.service

```
Before, After = 
Those two settings configure ordering dependencies between units. If unit foo.service contains the setting Before=bar.service and both units are being started, bar.service's start-up is delayed until foo.service has finished starting up. After= is the inverse of Before=, i.e. while Before= ensures that the configured unit is started before the listed unit begins starting up, After= ensures the opposite, that the listed unit is fully started up before the configured unit is started.
```

Then we would need to enable it on install of the microshift-networking-ovn package here:
https://github.com/openshift/microshift/blob/main/packaging/rpm/microshift.spec#L213

This would make systems wait on ovs init to finish before starting up microshift.

Reply:
But we still need to have them ordered, right?
How can we ensure that if the microshift-service doesn't now about `ovs-init`? Add `before: microshift` to the `ovs-init`?

Reply:
Alternative is we don't use systemd service for ovs init configuration and move it to CNI installation binary (not CNI binary). Just make sure the ovs configuration happens before actual CNI startup.